### PR TITLE
impersonate_ssl: added an SNI option for the ssl certificate request

### PR DESCRIPTION
**Pull request by friedrico**
This fix allows to set the server name indication for requesting the certificate
see:
https://github.com/rapid7/metasploit-framework/issues/14653#issuecomment-798945324
```
use auxiliary/gather/impersonate_ssl
set SNI www.google.com
set RHOSTS www.google.com
run
```

### CODE REVIEW

**Comment by friedrico** [comment]
I can not see why the Lint task failed. There is just an "error"

**Comment by gwillcox-r7** [comment]
```
--- Checking new and changed module syntax with tools/dev/msftidy.rb ---
modules/auxiliary/gather/impersonate_ssl.rb - [INFO] No CVE references found. Please check before you land!
modules/auxiliary/gather/impersonate_ssl.rb:46 - [WARNING] Please use snake case on method names: def getCert(rhost,rport,sni) 
```

**Comment by friedrico** [comment]
Thanks, I saw that one needs to view the raw log file in order to get the actual error message.
fixed that

**Comment by gwillcox-r7** [comment]
Applied some more RuboCop changes after running `rubocop -a modules/auxiliary/gather/impersonate_ssl.rb` to autofix a few errors and then I spent some time on fixing some of the others, as this module is really old and needed some touch ups. Also removed some dangerous `eval()` statements from the code that really didn't need to be there.

File: modules/auxiliary/gather/impersonate_ssl.rb [hunk 1]
```diff
@@ -42,8 +43,26 @@ def initialize(info={})
...
+      ctx = OpenSSL::SSL::SSLContext.new
+      sock = TCPSocket.new(rhost, rport)
+      ssl = OpenSSL::SSL::SSLSocket.new(sock, ctx)
+      if sni
+        ssl.hostname = sni
+      end
+      ssl.connect
+      cert = ssl.peer_cert
+      return cert
+  end
```
Comment (gwillcox-r7) [inline modules/auxiliary/gather/impersonate_ssl.rb]: So this, whilst nice, will break when pivoting comes into play. We should really be using the Rex socket methods and you can see that `peer_cert` is contained in https://github.com/rapid7/rex-socket/blob/86d71342aa8957fdc88f0075213b1cad8591ff92/lib/rex/socket/ssl_tcp.rb#L331-L333

**Review by gwillcox-r7 — DISMISSED** [review]
Overall looks good, but had one issue with the use of TCPSocket, see below for more info.

**Comment by gwillcox-r7** [comment]
Alright made some more changes and here are the results:

```
msf6 auxiliary(gather/impersonate_ssl) > show options

Module options (auxiliary/gather/impersonate_ssl):

   Name              Current Setting                   Required  Description
   ----              ---------------                   --------  -----------
   ADD_CN                                              no        Add CN to match spoofed site name (e.g. *.example.com)
   CA_CERT                                             no        CA Public certificate
   EXPIRATION                                          no        Date the new cert should expire (e.g. 06 May 2012, YESTERDAY or
                                                                 NOW)
   OUT_FORMAT        PEM                               yes       Output format (Accepted: DER, PEM)
   PRIVKEY                                             no        Sign the cert with your own CA private key
   PRIVKEY_PASSWORD                                    no        Password for private key specified in PRIV_KEY (if applicable)
   RHOSTS            www.google.com, www.facebook.com  yes       The target host(s), range CIDR identifier, or hosts file with sy
                                                                 ntax 'file:<path>'
   RPORT             443                               yes       The target port (TCP)
   SNI               www.google.com                    no        Server Name Indicator

msf6 auxiliary(gather/impersonate_ssl) > run
[*] Running module against 216.58.193.132

[*] 216.58.193.132:443 - Connecting to 216.58.193.132:443 SNI:www.google.com
[*] 216.58.193.132:443 - Copying certificate from 216.58.193.132:443
/C=US/ST=California/L=Mountain View/O=Google LLC/CN=www.google.com 
[*] 216.58.193.132:443 - Beginning export of certificate files
[*] 216.58.193.132:443 - Creating looted key/crt/pem files for 216.58.193.132:443
[+] 216.58.193.132:443 - key: /home/gwillcox/.msf4/loot/20210317172556_default_216.58.193.132_216.58.193.132_k_251470.key
[+] 216.58.193.132:443 - crt: /home/gwillcox/.msf4/loot/20210317172556_default_216.58.193.132_216.58.193.132_c_649836.crt
[+] 216.58.193.132:443 - pem: /home/gwillcox/.msf4/loot/20210317172556_default_216.58.193.132_216.58.193.132_p_138693.pem
[*] Running module against 2607:f8b0:4000:806::2004
[*] 2607:f8b0:4000:806::2004:443 - Connecting to 2607:f8b0:4000:806::2004:443 SNI:www.google.com
[*] 2607:f8b0:4000:806::2004:443 - Copying certificate from 2607:f8b0:4000:806::2004:443
/C=US/ST=California/L=Mountain View/O=Google LLC/CN=www.google.com 
[*] 2607:f8b0:4000:806::2004:443 - Beginning export of certificate files
[*] 2607:f8b0:4000:806::2004:443 - Creating looted key/crt/pem files for 2607:f8b0:4000:806::2004:443
[+] 2607:f8b0:4000:806::2004:443 - key: /home/gwillcox/.msf4/loot/20210317172556_default_2607f8b040008062004_2607f8b040008_498765.key
[+] 2607:f8b0:4000:806::2004:443 - crt: /home/gwillcox/.msf4/loot/20210317172556_default_2607f8b040008062004_2607f8b040008_611402.crt
[+] 2607:f8b0:4000:806::2004:443 - pem: /home/gwillcox/.msf4/loot/20210317172556_default_2607f8b040008062004_2607f8b040008_002246.pem
[*] Running module against 157.240.19.35
[*] 157.240.19.35:443 - Connecting to 157.240.19.35:443 SNI:www.google.com
[*] 157.240.19.35:443 - Copying certificate from 157.240.19.35:443
/C=US/ST=California/L=Mountain View/O=Google LLC/CN=www.google.com 
[*] 157.240.19.35:443 - Beginning export of certificate files
[*] 157.240.19.35:443 - Creating looted key/crt/pem files for 157.240.19.35:443
[+] 157.240.19.35:443 - key: /home/gwillcox/.msf4/loot/20210317172556_default_157.240.19.35_157.240.19.35_ke_391776.key
[+] 157.240.19.35:443 - crt: /home/gwillcox/.msf4/loot/20210317172556_default_157.240.19.35_157.240.19.35_ce_523569.crt
[+] 157.240.19.35:443 - pem: /home/gwillcox/.msf4/loot/20210317172556_default_157.240.19.35_157.240.19.35_pe_114485.pem
[*] Running module against 2a03:2880:f134:183:face:b00c:0:25de
[*] 2a03:2880:f134:183:face:b00c:0:25de:443 - Connecting to 2a03:2880:f134:183:face:b00c:0:25de:443 SNI:www.google.com
[*] 2a03:2880:f134:183:face:b00c:0:25de:443 - Copying certificate from 2a03:2880:f134:183:face:b00c:0:25de:443
/C=US/ST=California/L=Mountain View/O=Google LLC/CN=www.google.com 
[*] 2a03:2880:f134:183:face:b00c:0:25de:443 - Beginning export of certificate files
[*] 2a03:2880:f134:183:face:b00c:0:25de:443 - Creating looted key/crt/pem files for 2a03:2880:f134:183:face:b00c:0:25de:443
[+] 2a03:2880:f134:183:face:b00c:0:25de:443 - key: /home/gwillcox/.msf4/loot/20210317172557_default_2a032880f134183faceb00c025de_2a032880f1341_591335.key
[+] 2a03:2880:f134:183:face:b00c:0:25de:443 - crt: /home/gwillcox/.msf4/loot/20210317172557_default_2a032880f134183faceb00c025de_2a032880f1341_640475.crt
[+] 2a03:2880:f134:183:face:b00c:0:25de:443 - pem: /home/gwillcox/.msf4/loot/20210317172557_default_2a032880f134183faceb00c025de_2a032880f1341_396695.pem
[*] Auxiliary module execution completed
msf6 auxiliary(gather/impersonate_ssl) > 
```

**Comment by gwillcox-r7** [comment]
# Release Notes
The `impersonate_ssl.rb` module has been updated to add a new SNI option for retrieving the SSL Certificate, allowing it to properly retrieve SSL certificates in cases where the SNI option needs to be appropriately specified. In addition, RuboCop changes have also been applied to tidy up the code and remove some dangerous code in favor of safer solutions.

**Pull request merged**